ledger/ledger-common/src/main/kotlin/net/corda/v5/ledger/common/Amount.kt

```diff
@@ -0,0 +1,366 @@
+package net.corda.v5.ledger.common
+
+import net.corda.v5.base.annotations.CordaSerializable
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.util.Currency
+
+/**
+ * Represents an amount, which can be described as a positive quantity of an item.
+ *
+ * The quantity of an amount is represented as an integral value, as opposed to a decimal value.
+ *
+ * @param T The underlying type of the item being quantified by the current amount.
+ * @property quantity The quantity of the item of the current amount.
+ * @property item The item being quantified by the current amount.
+ * @property isDivisible Determines whether the current amount can be subdivided into smaller amounts.
+ */
+@CordaSerializable
+data class Amount<T : Any>(val quantity: BigInteger, val item: T) : Comparable<Amount<T>> {
```

Comment:
If we don't try to support to/from string representations and hence don't need to know scale and symbol then perhaps we don't need Amount. We just model it as BigInteger for the purposes of tracking fungible token counts.

Reply:
I guess on balance, if we've (customers have) misunderstood the `T` property, this should perhaps be the first thing to go?

With regards to handling precision, if we were to drop `T`, then `Amount` just becomes a useless wrapper for `BigInteger` and could be removed entirely, but if we do want a safer way to handle precision, then `Amount` sort of feels like we're just trying to re-implement `BigDecimal` with some adjusted behaviour, and I'm not sure we want to be doing that either.

The reason I didn't do anything with precision in the initial implementation was because I came to the realisation that ultimately you just end up dealing with the smallest indivisible unit (i.e. 1 penny/cent/satoshi, etc.) so just allow precision to be handled as a display/presentation issue. The second reason was that some instruments subdivide into different things, and therefore the precision changes based on how you want to interpret the thing you're looking at; for example:

1.00000000 BTC = 1 BTC or 0.00000001 MBTC (MegaBitcoin)
0.00000100 BTC = 1 uBTC or 10 Finney or 100 Satoshi

Reply:
Yes the C4 Amount was meant to allow easy conversion between display quantities and the indivisible tokens. It didn't really work that well as display is a complex topic. So it may well be we just use BigInteger. For me it is key that we look at the FungibleTokenState interface as that is where Amount is used. If that becomes: 
```
interface FungibleTokenState<T> :  ContractState {
   val tokenType: T
   val amount: BigInteger
   val issuer: Party // or PublicKey
   val owner: Party // or PublicKey
}
```

Then this does solve problems where issuer isn't a platform concept. Need to think about tokenType more as that is meant to represent different symbols/mixing rules, but that is left up to contract code